### ISSUE TITLE
Bootstrap Module build with Build.ps1 Script

### ISSUE THREAD

**Issue by gaelcolas** [PoshCode/ModuleBuilder#11]
For any PS Module project using ModuleBuilder, there will be a minimum code to be included in those repo to bootstrap the build process, namely the `Build.ps1` entry point.

It will have to at least:
- Boostrap the project by downloading dependencies such as ModuleBuilder if not already avail
- call `Build-Module`

That open the discussion to the following questions:

1. Shall we include a sample/re-usable version in ModuleBuilder, so that it can be used as part of a template?
2. How do we make the `Build.ps1` bootstrap process? 
```PowerShell
Save-Module ModuleBuilder
# Pull dependencies with PSDepend
Build-Module
```
Do we make a ModuleBuilder function call Invoke-PSDepend (in which case, do we make PSDepend a dependency of ModuleBuilder)?
3. Do we allow to change the PSModulePath as per #10 
4. Do we make `Build.ps1` also read the `Build.psd1` so that Configuration data (such as which Source to pull ModuleBuilder from during bootstrap)?
5. How do we customize the Build workflow (the steps to be called). Do we do this within Build.ps1, in external Data File? How do we manage the Interfaces for Build Steps (function signatures & Params)?

I need to refresh/refactor my SampleModule template, and I'd rather contribute to `ModuleBuilder` going forward but need 'guidance' as to how you envision solving those issues before I start submitting PRs.

**Comment by Jaykul** [comment]
I'm not sure I'm ready to make the decision about it. I'm not *completely* happy with *any* of these tools 😉 including this one 🤨 what do the rest of you think ("you" who are paying attention)?

At work, we're using home-built versions of all these tools, so I don't have a ton of experience with the various options. In fact, if you look at my older modules, they all have a big build.ps1 that handles dependencies and build tasks without relying on modules.

Do we want to try and make suggestions? 

Should we be doing that by example, or by building templates?

**Comment by gaelcolas** [comment]
I think we should start somewhere to allow criticism, and make it evolve over time.
The problem, that sort of sparkled this effort, is that everyone is building home-built versions of similar tools to solve similar issues with slightly unique variance of the same problem.
So best is to try something to find out what's wrong about it...

Let's not try to solve every problems at once, but release fast and release often, starting where we are with what we have... The community will make it the standard only if and when it's good enough.

I found that having a big `Build.ps1` is intimidating for many, and does not encourage change or sharing across projects. I feel that the `Build.ps1` should be the minimum to bootstrap and kick off the build, thus should be very standard, and every configuration of the build workflow should be done externally, if possible without requiring code (just the data), until you need specific/custom steps.

Finally, I don't think it matters whether we fail or not, as long as we (collectively) learn from the process (i.e. why it fails, and along agreed on some best practices through experiments).

**Comment by Jaykul** [comment]
I agree with all of that. Ultimately the issue with that is that you need a tool to bootstrap you with. In this case, the logical tool might be our own module -- which we would need to publish so we could use it?

**Comment by lipkau** [comment]
I currently have a `./Tools/setup.ps1` (like [this](https://github.com/AtlassianPS/JiraPS/blob/master/Tools/setup.ps1)) for bootstraping a clean agent/client.

The gist of what it does:
* # If PowerShellGet is not available (PSv4 and PSv3), it must be installed
* # Fail if PowerShellGet could not be found
* # PowerShell 5.1 and bellow need the PSGallery to be intialized
* # Make PSGallery trusted, to aviod a confirmation in the console
* # Installing PSDepend
* # Installing InvokeBuild
* # Installing Dependencies --> `Invoke-Build -task RunPSDepend`

And my PSDepend is currently pulling in the tarball of ModuleBuilder (which means I need addintion "post steps" to extract the package and build the module).

### TL;DR:
I would settle/prefer for ModuleBuilder to have a template of the bootstrapper, which I would copy and put in my root `./` or `./Tools` folder.
Preferably something that looks similar to other tools, eg: `MyModule.modulebuilder.ps1` or similar.

**Comment by Jaykul** [comment]
After trying repeatedly, I wasn't able to get PSDepend to work on Azure Pipelines, so I've come up with a bootstrap which does work:

https://github.com/PoshCode/ModuleBuilder/blob/34d68bdc0a20806b02c567a3498f7cef92409d67/bootstrap.ps1#L8-L12

For the sake of the example, I sort-of wanted to have it use PSDepends, but honestly, I'm happy with this for the build server -- I meant to set the Policy back at the end to avoid changing things on people's computers.

Thoughts?

I know you really like the notion of putting dependencies in a sub-folder, but it's quite complicated without a proper `venv` module (which doesn't exist yet, maybe we should write one). PSDepends can't seem to handle the situation for some reason, and the error is just not helpful (see the errors in [build 46](https://poshcode.visualstudio.com/ModuleBuilder/_build/results?buildId=46&view=logs))

**Comment by gaelcolas** [comment]
I haven't had a proper look at the error (I'm on mobile), but PSDepend is temperamental about the working directory iirc. You tried to set location in the folder containing psd1?
`venv` module would be soooo nice for DSC, but it's a mess, especially with some versions of Package Management...
That's a bit the idea with PSDepend, re-writing what needs to be would be great, but it's another Yak to shave...

**Comment by Jaykul** [comment]
Ok, I consider this done. There's a `bootstrap.ps1` to install things, and a `build.ps1` that builds ModuleBuilder using the code (and `azure-pipelines.yml` that makes it all happen on [azure pipelines](https://dev.azure.com/poshcode/ModuleBuilder/_build)) ...

For anything else open a new ticket 😉

**Comment by bravo-kernel** [comment]
Below, for future reference, a working PSDepend solution for Azure Pipelines.

<details>
<summary>azure-pipelines.yml</summary>

```yaml
- powershell: |
    Install-PackageProvider -Name NuGet -Force
    Install-Module PSDepend -Force
    Invoke-PSDepend -Path Requirements.psd1 -Target "CurrentUser" -Install -Import -Force -Verbose
  displayName: 'Install Required Modules'
```
</details>

<details>
<summary>Requirements.psd1</summary>

Used by the `-Path` parameter above.

```posh
@{
    PSDependOptions = @{
        # There is something to choose when it comes to `Target`:
        # - using scope `CurrentUser` will make sure the imported module will be available
        # in all following `powershell` pipeline tasks
        # - using a path (e.g. `.\Dependencies`) will install the modules into that path and 
        # while import will succeed the imports will only be available during that task
        # 
        # Even better, for best of both worlds we can override this setting with the `-Target`
        # parameter inside Azure Pipelines. This way local will use the path, azure the scope.
        Target = ".\Dependencies"

        Parameters = @{
            # Force should not be used inside Azure Pipelines (to prevent double parameter errors)
            #Force = $True
            SkipPublisherCheck = $true
        }
    }

    "Configuration" = "latest"
    "ModuleBuilder" = "latest"
    "Pester" = "latest"
    "PSScriptAnalyzer" = "latest"
}
```
</details>

**Comment by Jaykul** [comment]
That was neat, never seen the summary/details thing before, and I like the way you used the Target.
 
Does PSDepend have a way of specifying Min/Max for modules, such that we can say (since we know all of those modules follow SemVer), that we want the latest version, but without upgrades to new major version?

E.g. Configuration 1.* (but not 2.0)

**Comment by bravo-kernel** [comment]
Currently PSDepend only supports pinned/specific versions but there is a [PR waiting](https://github.com/RamblingCookieMonster/PSDepend/pull/96) to support min/max versions. Might be worth keeping an eye on that or even contributing there.

**Comment by gaelcolas** [comment]
That PR relies on something not that reliable, and I prefer how @Jaykul handles it in his bootstrap, but it has its own drawbacks. I'm not decided yet as how to proceed (and rather busy arm).

**Comment by bravo-kernel** [comment]
To prevent misunderstandings. This was added just as a future technical reference so it can be tested inside Azure Pipelines. You experts can best decide what solution will be the one. Either way, I will stick to that solution anyway since the module templates will be only my source of truth (once they become available).

**Comment by bravo-kernel** [comment]
FWIW I have now tested both solutions end-to-end and my vote goes to `Install-RequiredModule` 👍
